src/Desktop.Analyzers/Core/DoNotUseInsecureDTDProcessing.cs

```diff
@@ -31,85 +32,122 @@ public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
+                return s_supportedDiagnostics;
             }
         }
 
+        private void RegisterAnalyzer(OperationBlockStartAnalysisContext context, CompilationSecurityTypes types, Version frameworkVersion)
+        {
+            var analyzer = new Analyzer(types, frameworkVersion);
+            context.RegisterOperationAction(
+                analyzer.AnalyzeOperation,
+                OperationKind.InvocationExpression,
+                OperationKind.AssignmentExpression,
+                OperationKind.VariableDeclaration,
+                OperationKind.ObjectCreationExpression,
+                OperationKind.FieldInitializerAtDeclaration
+            );
+            context.RegisterOperationBlockEndAction(
+                analyzer.AnalyzeOperationBlock
+            );
+        }
+
+
         public override void Initialize(AnalysisContext analysisContext)
         {
             analysisContext.RegisterCompilationStartAction(
                 (context) =>
                 {
                     Compilation compilation = context.Compilation;
-                    var xmlTypes = new CompilationSecurityTypes(compilation);
-                    if (ReferencesAnyTargetType(xmlTypes))
+                    var _xmlTypes = new CompilationSecurityTypes(compilation);
+
+                    if (ReferencesAnyTargetType(_xmlTypes))
                     {
                         Version version = SecurityDiagnosticHelpers.GetDotNetFrameworkVersion(compilation);
-                        // bail if we are not analyzing project targeting .NET Framework
-                        // TODO: should we throw an exception to notify user?
+
                         if (version != null)
                         {
-                            context.RegisterCodeBlockStartAction<TLanguageKindEnum>(
+                            context.RegisterOperationBlockStartAction(
                                 (c) =>
                                 {
-                                    RegisterAnalyzer(c, xmlTypes, version);
+                                    RegisterAnalyzer(c, _xmlTypes, version);
                                 });
                         }
                     }
                 });
         }
 
-        private static bool ReferencesAnyTargetType(CompilationSecurityTypes types)
+        private class Analyzer
         {
-            return types.XmlDocument != null
-                || types.XmlNode != null
-                || types.XmlReader != null
-                || types.XmlTextReader != null
-                || types.XPathDocument != null
-                || types.XmlSchema != null
-                || types.DataSet != null
-                || types.DataTable != null
-                || types.DataViewManager != null
-                || types.XmlSerializer != null;
-        }
+            #region Environment classes
+            private class XmlDocumentEnvironment
```

Comment:
No. If structs, the object of the struct will get copied, instead of referenced by, when you pass it around thru functions, or put into dictionaries or get back from dictionaries. I definitely should avoid that.